crates/ra_hir/src/ty/infer.rs

```diff
@@ -583,20 +570,20 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                     .collect::<Vec<_>>()
                     .into();
 
-                Ty::Tuple(inner_tys)
+                Ty::apply(TypeName::Tuple, Substs(inner_tys))
             }
             Pat::Ref { pat, mutability } => {
-                let expectation = match *expected {
-                    Ty::Ref(ref sub_ty, exp_mut) => {
-                        if *mutability != exp_mut {
+                let expectation = match expected {
+                    Ty::Apply(ApplicationTy { name: TypeName::Ref(exp_mut), parameters }) => {
+                        if mutability != exp_mut {
                             // TODO: emit type error?
                         }
-                        &**sub_ty
+                        parameters.as_single()
```

Comment:
I wonder if we should add  an `&Ty -> Option<(Mutability, &Ty)>` methods that "casts" ty to `TypeName::Ref`?  That way, we'll encapsulate the panicy `as_single` call